tools/scripts/firehose.py

```diff
@@ -110,14 +119,39 @@ def cleanup(sql):
     conn.close()
 
 
-if __name__ == '__main__':
-    parser = argparse.ArgumentParser()
-    parser.add_argument('job')
-    parser.add_argument('--chunk', type=int, default=1000000000) # 1B by default
-    params = parser.parse_args()
-    chunk = params.chunk
-    print(datetime.datetime.utcnow().isoformat())
-
+def generate_jobs(jobs):
+    print(f'inserting {jobs} job(s)')
+    sys.path.insert(0, pkg_resources.get_distribution('awx').module_path)
+    from awx import prepare_env
+    prepare_env()
+    setup_django()
+
+    from awx.main.models import UnifiedJob, Job, JobTemplate
+
+    def make_batch(N):
+        jt = JobTemplate.objects.first()
+        jobs = [Job(job_template=jt, status='canceled', created=now(), modified=now(), elapsed=0.) for i in range(N)]
+        ujs = UnifiedJob.objects.bulk_create(jobs)
+        query = InsertQuery(Job)
+        fields = list(set(Job._meta.fields) - set(UnifiedJob._meta.fields))
```

Comment:
```python
for job in jobs:
  for field_name in fields:
    if not getattr(job, field_name) and getattr(jt, field_name):
      setattr(job, field_name)
```

you could save some computing cost by pre-creating a dict to update from the JT at the start, but this is my initial spitball of the concept

Reply:
Just realized there's more, also have to have special treatment for foreign keys so that you don't load them in the loop, which would defeat the point of the performance of the script.

Here's from when I was doing a similar thing copying JTs

```python
from time import time
from random import randint
from django.utils.timezone import now


jt_ct = JobTemplate.objects.count()


def copy_jts(N):
    s = time()
    print('start JT count {}'.format(JobTemplate.objects.count()))
    jts = JobTemplate.objects.order_by('created')[:jt_ct]
    for i in range(N):
        jt = jts[randint(0, jt_ct - 1)]
        data = jt.__dict__
        fields = set(f.name for f in JobTemplate._meta.get_fields() if f.editable)
        for fn in ('created_by', 'modified_by'):
            fields.add(fn)
        for key in list(data.keys()):
            if key.endswith('_id'):
                if key[:-len('_id')] not in fields and key not in fields:
                    data.pop(key)
            else:
                if key not in fields:
                    data.pop(key)
        for key in ('id', 'unifiedjobtemplate_ptr_id'):
            data.pop(key, None)
        for key in ('created', 'modified'):
            data[key] = now()
        JobTemplate.objects.create(**data)
    print('Make JTs took {} seconds, ct {}'.format(time() - s, JobTemplate.objects.count()))


# copy_jts(100)
```

That was working backwards from what you might do - taking all the data and then removing stuff not in fields, as opposed to adding things in fields. But it shows how the suffix `_id` needs special handling.

Reply:
🤷‍♂ the current code is simpler and it's pretty fast 🤷‍♂

Reply:
https://github.com/ansible/awx/pull/5973/files#r380692618

This `setattr` code actually seems to make each chunk about twice as slow for me.

Reply:
Without the related objects, things like this:

https://github.com/ansible/awx/pull/5726/files#diff-d8650fbe77b71f4895e7fa4ff18cc4d3R120-R129

won't cross over any tables. I don't know what use case this would be used for, but it would be meaningless for the use case that I have in that PR. Twice as slow doesn't sound bad to me for linking projects.

It also seems likely that it won't be twice as slow if it's passed into the `create` as opposed to a latter setattr, which should also be logically possible.